// @wesdigital/adv-cdk-modules/dist/ecs-fargate/load-balanced-service.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LoadBalancedService = void 0;
const constructs_1 = require("constructs");
const aws_ecs_patterns_1 = require("aws-cdk-lib/aws-ecs-patterns");
const constructs_2 = require("./constructs");
class LoadBalancedService extends constructs_1.Construct {
    constructor(scope, id, props) {
        super(scope, id);
        const serviceDefinition = new constructs_2.CommonServiceDef(this, 'Definition', {
            fargate: props,
            serviceType: aws_ecs_patterns_1.ApplicationLoadBalancedFargateService
        });
        this._kmsKey = serviceDefinition.kmsKey;
        this._service = serviceDefinition.service;
        this._service.targetGroup.configureHealthCheck(props.albHealthCheck);
    }
    get service() {
        return this._service;
    }
    get loadBalancer() {
        return this._service.loadBalancer;
    }
    get kmsKey() {
        return this._kmsKey;
    }
}
exports.LoadBalancedService = LoadBalancedService;
