// @wesdigital/adv-cdk-modules/dist/eventbridge/bus.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ArchivedEventBridgeBus = void 0;
const constructs_1 = require("constructs");
const aws_events_1 = require("aws-cdk-lib/aws-events");
const aws_cdk_lib_1 = require("aws-cdk-lib");
class ArchivedEventBridgeBus extends constructs_1.Construct {
    constructor(scope, id, props) {
        super(scope, id);
        const eventBusName = `${props.deploymentEnv}-${props.busName}`;
        const bus = new aws_events_1.EventBus(this, 'EventBus', {
            eventBusName
        });
        this._bus = bus;
        bus.archive('Archive', {
            archiveName: `${eventBusName}-archive-all`,
            description: `Archives all events from ${eventBusName}`,
            retention: aws_cdk_lib_1.Duration.days(props.archiveRetentionInDays),
            eventPattern: {
                account: [aws_cdk_lib_1.Stack.of(this).account]
            }
        });
    }
    get bus() {
        return this._bus;
    }
}
exports.ArchivedEventBridgeBus = ArchivedEventBridgeBus;
